crates/noirc_evaluator/src/ssa/opt/mem2reg.rs

```diff
@@ -122,21 +150,117 @@ impl PerBlockContext {
         // to read.
-        dfg[self.block_id]
+        dfg[block_id]
             .instructions_mut()
             .retain(|instruction| !loads_to_substitute.contains_key(instruction));
 
         protected_allocations
     }
 
+    // This method finds the load values to substitute for a given address
+    // The search starts at the block supplied as a parameter. If there is not a load to substitute
+    // the CFG is analyzed to determine whether a predecessor block has a load value to substitute.
+    // If there is no load value to substitute the original address is returned.
+    fn fetch_load_value_to_substitute(
+        &mut self,
+        block_id: BasicBlockId,
+        address: ValueId,
+    ) -> ValueId {
+        let mut stack = vec![block_id];
+        let mut visited = HashSet::new();
+
+        while let Some(block) = stack.pop() {
+            visited.insert(block);
+
+            // Check whether there is a load value to substitute in the current block.
+            // Return the value if found.
+            if let Some((value, load_block_id)) =
+                self.load_values_to_substitute_per_func.get(&address)
+            {
+                if *load_block_id == block {
+                    return *value;
+                }
+            }
+
+            // If no load values to substitute have been found in the current block, check the block's predecessors.
+            let predecessors = self.cfg.predecessors(block);
+            for predecessor in predecessors {
+                if self.dom_tree.is_reachable(predecessor)
+                    && self.dom_tree.dominates(predecessor, block)
+                    && !visited.contains(&predecessor)
+                {
+                    stack.push(predecessor);
+                }
+            }
+        }
+        address
+    }
+
+    // This method determines which loads should be substituted.
+    // Starting at the block supplied as a parameter, we check whether a store has occurred with the given address.
+    // If no store has occurred in the supplied block, the CFG is analyzed to determine whether
+    // a predecessor block has a store at the given address.
+    fn find_load_to_substitute(
+        &mut self,
+        block_id: BasicBlockId,
+        address: ValueId,
+        dfg: &DataFlowGraph,
+        instruction_id: &InstructionId,
+        loads_to_substitute: &mut BTreeMap<InstructionId, ValueId>,
+        load_values_to_substitute_per_block: &mut BTreeMap<ValueId, ValueId>,
+    ) -> bool {
+        let mut stack = vec![block_id];
+        let mut visited = HashSet::new();
+
+        while let Some(block) = stack.pop() {
+            visited.insert(block);
+
+            for l in self.loops.yet_to_unroll.iter() {
+                // We do not want to substitute loads that take place within loops as this pass
+                // can occur before loop unrolling
+                if block == l.header {
+                    return false;
+                }
```

Comment:
If this pass can't handle loads within loops, will it also break if a load is used after a loop which mutates it?
```rs
fn main() {
    let ref = &mut 0;
    for _ in 0 .. 10 {
        *ref += 1;
    }
    assert(*ref == 10);
}
```
Or variations of the above?

Reply:
It would break if we also removed the mem2reg pass that occurs after flattening. This snippet passes as we run mem2reg before loop unrolling and also after flattening. I have actually updated this as I found bug while working on merge slices which is based off of this PR. I no longer construct a `Loops` object and instead only continue searching for stores on blocks with one predecessor. This enables the mem2reg to actually be fully re-orderable because if we remove loads/stores in blocks with multiple predecessors we risk overwriting other comptime values that have been simplified in other passes such as loop unrolling or flattening. 

For example this snippet:
```
    let mut new_slice = [];
    for i in 0..5 {
        new_slice = new_slice.push_back(i);
    }
    assert(new_slice.len() == 5);
```
will provide this SSA after inlining:
```
After Inlining:
acir fn main f1 {
  b0():
    v0 = allocate
    store [] at v0
    jmp b1(Field 0)
  b1(v2: Field):
    v5 = lt v2, Field 5
    jmpif v5 then: b2, else: b3
  b2():
    v10 = load v0
    v12 = call slice_push_back(v10, v2)
    store v12 at v0
    v14 = add v2, Field 1
    jmp b1(v14)
  b3():
    v6 = load v0
    v8 = call array_len(v6)
    v9 = eq v8, Field 5
    constrain v9
    return 
}
```
If we were load v0 in block 2, the first argument to `slice_push_back` would be replaced with the constant `[]`. 
Bad SSA:
```
acir fn main f1 {
  b0():
    v0 = allocate
    jmp b1(Field 0)
  b1(v2: Field):
    v5 = lt v2, Field 5
    jmpif v5 then: b2, else: b3
  b2():
    v12 = call slice_push_back([], v2)
    v14 = add v2, Field 1
    jmp b1(v14)
  b3():
    v8 = call array_len([])
    v9 = eq v8, Field 5
    constrain v9
    return 
}
```
While in fact we should only remove the load once loop unrolling is done and there are five `slice_push_back` calls, so that only the first value is replaced.